### PR TITLE
[SPARK-48615][SQL] Perf improvement for parsing hex string

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

Currently, we use two heximal string parsing functions. One uses Apache Codecs Hex for X-prefixed lit parsing, and the other use builtin impl for unhex function. I did a benchmark for them comparing with the `java.util.HexFormat` which was introduced in JDK17.

```
OpenJDK 64-Bit Server VM 17.0.10+0 on Mac OS X 14.5
Apple M2 Max
Cardinality 1000000:                      Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Apache                                             5050           5100          86          0.2        5050.1       1.0X
Spark                                              3822           3840          30          0.3        3821.6       1.3X
Java                                               2462           2522          87          0.4        2462.1       2.1X

OpenJDK 64-Bit Server VM 17.0.10+0 on Mac OS X 14.5
Apple M2 Max
Cardinality 2000000:                      Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Apache                                            10020          10828        1154          0.2        5010.1       1.0X
Spark                                              6875           6966         144          0.3        3437.7       1.5X
Java                                               4999           5092          89          0.4        2499.3       2.0X

OpenJDK 64-Bit Server VM 17.0.10+0 on Mac OS X 14.5
Apple M2 Max
Cardinality 4000000:                      Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Apache                                            20090          20433         433          0.2        5022.5       1.0X
Spark                                             13389          13620         229          0.3        3347.2       1.5X
Java                                              10023          10069          42          0.4        2505.6       2.0X

OpenJDK 64-Bit Server VM 17.0.10+0 on Mac OS X 14.5
Apple M2 Max
Cardinality 8000000:                      Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Apache                                            40277          43453        2755          0.2        5034.7       1.0X
Spark                                             27145          27380         311          0.3        3393.1       1.5X
Java                                              19980          21198        1473          0.4        2497.5       2.0X
```

The results indicate that the speed is Apache Codecs < builtin < Java, increasing by ~50%.

In this PR, we replace these two with the Java 17 API

### Why are the changes needed?

performance enhance

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

benchmarking 

existing unit tests in org.apache.spark.sql.catalyst.expressions.MathExpressionsSuite

### Was this patch authored or co-authored using generative AI tooling?

no

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan @dongjoon-hyun @LuciferYang thanks

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala [hunk 1]
```diff
@@ -1162,41 +1140,27 @@ case class Unhex(child: Expression, failOnError: Boolean = false)
...
-    if (failOnError && result == null) {
-      // The failOnError is set only from `ToBinary` function - hence we might safely set `hint`
-      // parameter to `try_to_binary`.
-      throw QueryExecutionErrors.invalidInputInConversionError(
-        BinaryType,
-        num.asInstanceOf[UTF8String],
-        UTF8String.fromString("HEX"),
-        "try_to_binary")
+    try {
+      Hex.unhex(num.asInstanceOf[UTF8String].toString)
```
Comment (LuciferYang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:1144]: The improvement to the `unhex` function seems effective in isolated cases, but IIRC, `UTF8String.toString` always incurs an additional Array Copy overhead compared to `UTF8String.getBytes`. Is this really better? 

Therefore, should we add a offline test to evaluate the effectiveness of the improvement for the `Unhex` sql function, possibly focusing on both short and long strings?

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:1144]: Good point, let me implement unhex directly on bytes

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 2]
```diff
@@ -2788,11 +2786,10 @@ class AstBuilder extends DataTypeAstBuilder with SQLConfHelper with Logging {
           Literal(interval, CalendarIntervalType)
         }
       case BINARY_HEX =>
-        val padding = if (value.length % 2 != 0) "0" else ""
```
Comment (yaooqinn) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2791]: We needn't padding anymore, as we can handle odd length values

**Review by LuciferYang — APPROVED** [review]

**Comment by LuciferYang** [comment]
Merged into master for Spark 4.0. Thanks @yaooqinn

**Comment by panbingkun** [comment]
I found that the performance of `hex(Array[Byte])` using JDK seems to be better than `Spark build-in method`. @yaooqinn, do you think this can be done?

**Comment by panbingkun** [comment]
OpenJDK 64-Bit Server VM 17.0.10+7-LTS on Mac OS X 14.5
Apple M2
Cardinality 1000000:                      Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Common Codecs                                      2632           2647          23          0.4        2632.4       1.0X
Java                                               2160           2197          60          0.5        2159.8       1.2X
Spark                                              8675           8848         223          0.1        8675.4       0.3X

**Comment by yaooqinn** [comment]
Hi @panbingkun, that is on my schedule. But if you are interested in this improvement, you can take it over. This is an umbrella ticket, and you can include a subtask there https://issues.apache.org/jira/browse/SPARK-48624

**Comment by panbingkun** [comment]
> Hi @panbingkun, that is on my schedule. But if you are interested in this improvement, you can take it over. This is an umbrella ticket, and you can include a subtask there https://issues.apache.org/jira/browse/SPARK-48624

Okay, actually I just wanted to ask if you have started, to `avoid conflicts`. If you have started, please continue, and I will take a look at the others. Thanks!

**Comment by panbingkun** [comment]
https://issues.apache.org/jira/browse/SPARK-48640
You reuse this jira, Please add it to above umbrella ticket. I don't have the authority to operate this, 😄

**Comment by yaooqinn** [comment]
Hi @panbingkun， Based on my benchmarking result, the Spark version of `hex(Array[Byte])` appears to outperform the one from JDK. You can send your PR, if you do get performance gain while switching to JDK

**Comment by panbingkun** [comment]
> Hi @panbingkun， Based on my benchmarking result, the Spark version of `hex(Array[Byte])` appears to outperform the one from JDK. You can send your PR, if you do get performance gain while switching to JDK

Okay.

**Pull request closed without merging**